### PR TITLE
[FIX] *: strict api for main orm methods

### PR DESCRIPTION
**Pull request by Feyensv**
Enforce strict types for returned values for
* create
* write
* unlink
* default_get

to make those methods more consistent and reliable. Also make sure they can be called with empty self/values, i.e. that they follow the same behavior as the base methods defined in the main orm Model.

Enterprise PR: https://github.com/odoo/enterprise/pull/38880

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_tax.py [hunk 1]
```diff
@@ -209,18 +209,21 @@ def _compute_tax_group_id(self):
...
-
-        repartition = rslt.setdefault('repartition_line_ids', [])
-        if 'repartition_line_ids' in fields_list and not repartition:
-            repartition.extend([
+        if 'company_id' not in fields_list and (
+            set(['refund_repartition_line_ids', 'invoice_repartition_line_ids'])
+            & set(fields_list)
+        ):
+            fields_list += ['company_id']
+        rslt = super(AccountTax, self).default_get(fields_list)
```
Comment (Julien00859) [inline addons/account/models/account_tax.py:217]: Nitpicking, fell free to ignore

* literal-set over `set([...])`
* `not set().isdisjoint(iterable)` is faster and uses less memory than `set() & set()`
* if I had to change a single thing in Python, I would remove `list.__iadd__`, this function really is misleading, I always prefer append/extend
* py3 super
```suggestion
        if 'company_id' not in fields_list and (
            not {'refund_repartition_line_ids', 'invoice_repartition_line_ids'}.isdisjoint(fields_list)
        ):
            fields_list.append('company_id')
        rslt = super().default_get(fields_list)
```

File: addons/account/models/ir_module.py [hunk 2]
```diff
@@ -61,13 +61,14 @@ def _compute_account_templates(self):
     def write(self, vals):
         # Instanciate the first template of the module on the current company upon installing the module
         was_installed = len(self) == 1 and self.state == 'installed'
-        super().write(vals)
+        res = super().write(vals)
         is_installed = len(self) == 1 and self.state == 'installed'
-        if not was_installed and is_installed and not self.env.company.chart_template and self.account_templates:
+        if len(self) == 1 and not was_installed and is_installed and not self.env.company.chart_template and self.account_templates:
```
Comment (Julien00859) [inline addons/account/models/ir_module.py:66]:
```suggestion
        if not was_installed and is_installed and not self.env.company.chart_template and self.account_templates:
```
part of both `was_installed` and `is_installed` already

File: addons/crm/wizard/crm_lead_to_opportunity.py [hunk 3]
```diff
@@ -16,7 +16,10 @@ def default_get(self, fields):
         to ease window action definitions, and be backward compatible. """
         result = super(Lead2OpportunityPartner, self).default_get(fields)
 
-        if not result.get('lead_id') and self.env.context.get('active_id'):
+        if 'lead_id' in fields and not result.get('lead_id') and self.env.context.get('active_id'):
+            lead = self.env['crm.lead'].browse(self.env.context.get('active_id'))
+            if lead.probability == 100:
+                raise UserError(_("Closed/Dead leads cannot be converted into opportunities."))
```
Comment (Julien00859) [inline addons/crm/wizard/crm_lead_to_opportunity.py:22]: what is this? seems unrelated

File: addons/mail_group/models/mail_group.py [hunk 4]
```diff
@@ -39,7 +39,7 @@ class MailGroup(models.Model):
     @api.model
     def default_get(self, fields):
         res = super(MailGroup, self).default_get(fields)
-        if not res.get('alias_contact') and (not fields or 'alias_contact' in fields):
+        if not res.get('alias_contact') and 'alias_contact' in fields:
```
Comment (Julien00859) [inline addons/mail_group/models/mail_group.py:42]: Nitpicking, feel free to ignore
```suggestion
        if 'alias_contact' in fields and not res['alias_contact']:
```

File: addons/point_of_sale/data/point_of_sale_onboarding.xml [hunk 5]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
-<odoo>
+<odoo noupdate="1">
```
Comment (Julien00859) [inline addons/point_of_sale/data/point_of_sale_onboarding.xml:2]: seems unrelateed

File: addons/project/models/project_task.py [hunk 6]
```diff
@@ -807,7 +807,7 @@ def default_get(self, default_fields):
             project = self.env['project.project'].browse(project_id)
             if project.analytic_account_id:
                 vals['analytic_account_id'] = project.analytic_account_id.id
-        elif 'default_user_ids' not in self.env.context:
+        elif 'default_user_ids' not in self.env.context and 'user_ids' in default_fields:
             user_ids = vals.get('user_ids', [])
             user_ids.append(Command.link(self.env.user.id))
             vals['user_ids'] = user_ids
```
Comment (Julien00859) [inline addons/project/models/project_task.py:813]: Nitpicking, feel free to ignore
```suggestion
            vals['user_ids'].append(Command.link(self.env.user.id))
```

File: addons/sms/models/sms_template.py [hunk 7]
```diff
@@ -15,7 +15,7 @@ class SMSTemplate(models.Model):
     @api.model
     def default_get(self, fields):
         res = super(SMSTemplate, self).default_get(fields)
```
Comment (Julien00859) [inline addons/sms/models/sms_template.py:17]: Since we're at it
```suggestion
        res = super().default_get(fields)
```

File: odoo/addons/base/models/res_config.py [hunk 8]
```diff
@@ -466,12 +466,14 @@ def get_values(self):
 
     @api.model
     def default_get(self, fields):
+        res = super(ResConfigSettings, self).default_get(fields)
```
Comment (Julien00859) [inline odoo/addons/base/models/res_config.py:469]: Since we're at it
```suggestion
        res = super().default_get(fields)
```

File: odoo/addons/base/tests/test_overrides.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+            model_env = self.env[model]
+            if model_env._transient:
+                continue
+            try:
+                # with self.assertQueryCount(1):  # allow one query for the call to get_model_defaults.
+                self.assertEqual(
+                    model_env.browse().default_get([]), {},
+                    "Invalid default_get return value for model %s" % model_env._name)
+            except UserError:
+                # skip "You must be logged in a Belgian company to use this feature" errors
```
Comment (Julien00859) [inline odoo/addons/base/tests/test_overrides.py:48]: best if you could skip the model that would raise this error

File: odoo/addons/base/tests/test_overrides.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+            model_env = self.env[model]
+            if model_env._abstract:
+                continue
+            try:
+                # with self.assertQueryCount(0):
+                self.assertEqual(
+                    model_env.browse().write({}), True,
+                    "Invalid write return value for model %s" % model_env._name)
+            except UserError:
+                # skip models that should never be modified
```
Comment (Julien00859) [inline odoo/addons/base/tests/test_overrides.py:34]: best if you could make a list of "to be skipped" modules and to skip them instead of try/except

**Review by Julien00859 — APPROVED** [review]
A few nitpickings, I don't think I spotted any error, didn't read enterprise yet

For the unrelated things, I'm fine as long as they are in their own commits

File: addons/account/models/ir_module.py (hunk 2 above)
Comment (Feyensv) [reply addons/account/models/ir_module.py:66]: Not really, it'll try loading the chart template and crash when self is empty

File: addons/crm/wizard/crm_lead_to_opportunity.py (hunk 3 above)
Comment (Feyensv) [reply addons/crm/wizard/crm_lead_to_opportunity.py:22]: Oh, probably a rebase conflict :eyes:

File: odoo/addons/base/tests/test_overrides.py (hunk 10 above)
Comment (Feyensv) [reply odoo/addons/base/tests/test_overrides.py:34]: Well, there is a bunch of them, and I don't want to force people to modify this list here and there.
UserError are custom Odoo errors, so I believe discarding those ones for now is acceptable :thinking:

File: odoo/addons/base/tests/test_overrides.py (hunk 9 above)
Comment (Feyensv) [reply odoo/addons/base/tests/test_overrides.py:48]: same answer

File: addons/project/models/project_task.py (hunk 6 above)
Comment (Feyensv) [reply addons/project/models/project_task.py:813]: Why? there is no guarantee that the returned dict of `defaut_get` holds a value for each field.
`vals['user_id']` could crash :thinking:

File: addons/point_of_sale/data/point_of_sale_onboarding.xml (hunk 5 above)
Comment (Feyensv) [reply addons/point_of_sale/data/point_of_sale_onboarding.xml:2]: had an error due to this, trying to remove it.

File: addons/project/models/project_task.py (hunk 6 above)
Comment (Julien00859) [reply addons/project/models/project_task.py:787]: really? damn I'm a noob x)

File: addons/account/models/ir_module.py (hunk 2 above)
Comment (Julien00859) [reply addons/account/models/ir_module.py]: I don't get it, can you elaborate?

File: addons/point_of_sale/models/pos_session.py [hunk 11]
```diff
@@ -213,10 +214,11 @@ def create(self, vals_list):
                 'update_stock_at_closing': update_stock_at_closing,
             })
 
+        self_ctxt = self.with_context(**context)
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/point_of_sale/models/pos_session.py:217]: Yeah, it's quite crappy, I tried to avoid modifying the existing logic :woman_shrugging:

File: odoo/addons/base/tests/test_overrides.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.exceptions import UserError
+from odoo.tests import TransactionCase, tagged
+
+
+@tagged('-at_install', 'post_install')
+class TestOverrides(TransactionCase):
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply odoo/addons/base/tests/test_overrides.py:8]: Indeed it doesn't ensure the full coherence with a different loading order, doesn't me that it's not helpful to check that much :thinking:.

**Comment by Feyensv** [comment]
@robodoo r+

**Comment by MiquelRForgeFlow** [comment]
@Feyensv FYI you could also have considered copy() method

**Comment by Feyensv** [comment]
> @Feyensv FYI you could also have considered copy() method

Didn't think of it, but `copy` can only be called with a record and therefore cannot be magically handled in one test for all models.
